packages/workbox-routing/lib/Route.mjs

```diff
@@ -0,0 +1,170 @@
+ * undefined if nothing was returned.
+ * @return {Promise<Response>} The response that will fulfill the request.
+ * @memberof module:workbox-routing
+ */
+
+/**
+ * A `Route` allows you to tell a service worker that it should handle
+ * certain network requests using a specific response strategy.
+ *
+ * A consists or a matcher and a handler. A matcher needs to determine if a
+ * route should be used for a request. A handler should handle the request
+ * if it does match a Router.
+ *
+ * Instead of implementing your own handlers, you can use one of the
+ * pre-defined runtime caching strategies from the
+ * {@link module:workbox-runtime-caching|workbox-runtime-caching} module.
+ *
+ * There are also pre-defined Route's provided by this library:
+ * {@link module:workbox-routing.RegExpRoute|RegExpRoute}
+ * and {@link module:workbox-routing.ExpressRoute|ExpressRoute} subclasses
+ * which provide a convenient wrapper with a nicer interface for using regular
+ * expressions or Express-style routes as the `match` criteria.
+ *
+ * @example
+ * // Any navigate requests for URLs that start with /path/to/ will match.
+ * const route = new workbox.routing.Route({
+ *   match: ({url, event}) => {
+ *     return event.request.mode === 'navigate' &&
+ *            url.pathname.startsWith('/path/to/');
+ *   },
+ *   handler: ({event}) => {
+ *     // Do something that returns a Promise<Response>, like:
+ *     return caches.match(event.request);
+ *   },
+ * });
+ *
+ * const router = new workbox.routing.Router();
+ * router.registerRoute({route});
+ *
+ * @memberof module:workbox-routing
+ */
+export default class Route {
+  /**
+   * Constructor for Route class.
+   * @param {Object} input
+   * @param {function} input.match The function that determines whether the
+   * route matches a given `fetch` event.
+   *
+   * See [matchCallback]{@link module:workbox-routing.Route~matchCallback} for
+   * full details on this function.
+   * @param {function|module:workbox-runtime-caching.Handler} input.handler
+   * This parameter can be either a function or an object which is a subclass
+   * of `Handler`.
+   *
+   * Either option should result in a `Response` that the `Route` can use to
+   * handle the `fetch` event.
+   *
+   * See [handlerCallback]{@link module:workbox-routing.Route~handlerCallback}
+   * for full details on using a callback function as the `handler`.
+   * @param {string} [input.method] Only match requests that use this
+   * HTTP method.
+   *
+   * Defaults to `'GET'`.
+   */
+  constructor({match, handler, method} = {}) {
+    this.handler = Route._normalizeHandler(handler);
+
+    if (process.env.NODE_ENV !== 'production') {
+      core.assert.isType(match, 'function', {
+        moduleName: 'workbox-routing',
+        funcName: 'Route',
+        paramName: 'match',
+      });
+    }
+    this.match = match;
+
+    if (method) {
+      if (process.env.NODE_ENV !== 'production') {
```

Comment:
Nit: Could you move this in with the previous if statement? I know it'll require a second if on the method, but would at least group the assertions and make it clear what code is stripped.

Reply:
Okay, I've rearranged things to group together the assertions.